File: applications/event-account-http-source/src/main/java/showcase/event/stream/rabbitmq/account/http/source/RabbitStreamConfig.java
package showcase.event.stream.rabbitmq.account.http.source;

import com.rabbitmq.stream.Environment;
import com.rabbitmq.stream.Producer;
import com.vmware.tanzu.data.services.rabbitmq.streaming.account.domain.Account;
import lombok.extern.slf4j.Slf4j;
import nyla.solutions.core.patterns.conversion.Converter;
import nyla.solutions.core.patterns.integration.Publisher;
import org.springframework.amqp.core.Queue;
import org.springframework.amqp.core.QueueBuilder;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Profile;
import org.springframework.rabbit.stream.producer.RabbitStreamTemplate;

import static java.lang.String.valueOf;

@Configuration
@Slf4j
@Profile("stream")
public class RabbitStreamConfig {

    private static final String FILTER_PROP_NM = "stateProvince";

    @Value("${spring.cloud.stream.bindings.output.destination}")
    private String streamName;

    @Value("${rabbitmq.streaming.use.filter:false}")
    private boolean isUseFilter;

    @Bean
    Queue stream(Environment environment) {
        log.info("Creating stream: {}",streamName);

        environment.streamCreator().name(streamName)
                .create();

        return QueueBuilder.durable(streamName)
                .build();
    }

    @Bean
    Producer producer(Environment environment)
    {
        log.info("stream: {}, isUseFilter: {}",streamName,isUseFilter);
        var builder = environment.producerBuilder()
                .stream(streamName);

        if(isUseFilter){
               builder = builder.filterValue(msg ->
                    valueOf(msg.getApplicationProperties().get(FILTER_PROP_NM)));
        }

        return builder.build();
    }

    @Bean
    Publisher<Account> publisher(Producer producer, Converter<Account,byte[]> converter)
    {
        return account ->{

            var state = account.getLocation() != null ? account.getLocation().getStateProvince() : "";

            producer.send(producer.messageBuilder()
                    .addData(converter.convert(account))
                    .properties().messageId(account.getId())
                    .messageBuilder()
                    .applicationProperties().entry(FILTER_PROP_NM,state)
                    .messageBuilder()
                    .build(), confirmationStatus ->{});
        };
    }

    @Bean
    RabbitStreamTemplate rabbitStreamTemplate(Environment environment) {
        return new RabbitStreamTemplate(environment,streamName);
    }

}

File: applications/sources/account-generator-streams-source/src/main/kotlin/com/vmware/tanzu/data/services/rabbitmq/streaming/account/RabbitStreamConfig.kt
package com.vmware.tanzu.data.services.rabbitmq.streaming.account

import com.rabbitmq.stream.*
import org.springframework.amqp.rabbit.listener.RabbitListenerContainerFactory
import org.springframework.amqp.support.converter.Jackson2JsonMessageConverter
import org.springframework.amqp.support.converter.MessageConverter
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Primary
import org.springframework.rabbit.stream.config.StreamRabbitListenerContainerFactory
import org.springframework.rabbit.stream.listener.ConsumerCustomizer
import org.springframework.rabbit.stream.listener.StreamListenerContainer
import org.springframework.rabbit.stream.producer.RabbitStreamTemplate


/**
 *
 */
@Configuration
class RabbitStreamConfig() {

    @Value("\${spring.rabbitmq.username:guest}")
    private var username: String = "guest";

    @Value("\${account.generator.stream.name:event-streaming-showcase.accounts}")
    private val streamName: String = "event-streaming-showcase.accounts"

    @Value("\${spring.rabbitmq.password:guest}")
    private var password:  String = "guest";

    @Value("\${spring.rabbitmq.host:localhost}")
    private var hostname: String = "localhost";

    @Value("\${spring.application.name}")
    private var applicationName: String? = null


    @Bean
    fun rabbitStreamEnvironment(): Environment {

        val env =  Environment.builder()
            .host(hostname)
            .username(username)
            .password(password).build()


//       try{
//           env.streamCreator().stream(streamName).create()
//       }
//       catch(e : StreamException)
//       {
//       }

        return env

    }

    @Bean
    fun streamTemplate(env: Environment): RabbitStreamTemplate {
        val template = RabbitStreamTemplate(env, streamName)
        template.setProducerCustomizer { name: String?, builder: ProducerBuilder ->
            builder.name(
                "test"
            )
        }

        template.setMessageConverter(messageConverter())

        return template
    }

//    @Bean
//    fun nativeFactory( env : Environment)  : RabbitListenerContainerFactory<StreamListenerContainer> {
//        var factory =  StreamRabbitListenerContainerFactory(env);
//        factory.setNativeListener(true);
//
//
////        env.streamCreator().stream(streamName).
//
//        val cc : ConsumerCustomizer = ConsumerCustomizer{ id: String, builder: ConsumerBuilder ->
//            builder.name("myConsumer")
//                .offset(OffsetSpecification.first())
//                .manualTrackingStrategy();
//        }
//        factory.setConsumerCustomizer(cc);
//        return factory;
//    }

    @Primary
    @Bean
    fun messageConverter() : MessageConverter
    {
        return Jackson2JsonMessageConverter();
    }

//    @RabbitListener(queues = ["test.stream.queue1"],containerFactory = "nativeFactory")
//    fun streamListener(input : String) {
//    }
//
//    @RabbitListener(id = "test", queues = ["test.stream.queue2"], containerFactory = "nativeFactory")
//    void nativeMsg(Message in, Context context) {
//        ...
//        context.storeOffset();
//    }

}